src/sentry/tasks/assemble.py

```diff
@@ -284,12 +285,66 @@ def _extract_debug_ids_from_manifest(
     return bundle_id, debug_ids_with_types
 
 
-def _remove_duplicate_artifact_bundles(bundle: ArtifactBundle, bundle_id: str):
+def _remove_duplicate_artifact_bundles(org_id: int, ids: List[int]):
+    # In case there are no ids to delete, we don't want to run the query, otherwise it will result in a deletion of
+    # all ArtifactBundle(s) with the specific bundle_id.
+    if not ids:
+        return
+
     # Even though we delete via a QuerySet the associated file is also deleted, because django will still
     # fire the on_delete signal.
-    ArtifactBundle.objects.filter(
-        ~Q(id=bundle.id), bundle_id=bundle_id, organization_id=bundle.organization_id
-    ).delete()
+    ArtifactBundle.objects.filter(Q(id__in=ids), organization_id=org_id).delete()
+
+
+def _bind_or_create_artifact_bundle(
+    bundle_id: uuid,
+    date_added: datetime,
+    org_id: int,
+    archive_file: File,
+    artifact_count: int,
+) -> Tuple[ArtifactBundle, bool]:
+    existing_artifact_bundles = list(
+        ArtifactBundle.objects.filter(organization_id=org_id, bundle_id=bundle_id)
+    )
+
+    if len(existing_artifact_bundles) == 0:
+        existing_artifact_bundle = None
+    else:
+        existing_artifact_bundle = existing_artifact_bundles.pop()
+        # We want to remove all the duplicate artifact bundles that have the same bundle_id.
+        _remove_duplicate_artifact_bundles(
+            org_id=org_id, ids=list(map(lambda value: value.id, existing_artifact_bundles))
+        )
+
+    # In case there is not ArtifactBundle with a specific bundle_id, we just create it and return.
+    if existing_artifact_bundle is None:
+        artifact_bundle = ArtifactBundle.objects.create(
+            organization_id=org_id,
+            # In case we didn't find the bundle_id in the manifest, we will just generate our own.
+            bundle_id=bundle_id or uuid.uuid4().hex,
+            file=archive_file,
+            artifact_count=artifact_count,
+            # For now these two fields will have the same value but in separate tasks we will update "date_added"
+            # in order to perform partitions rebalancing in the database.
+            date_added=date_added,
+            date_uploaded=date_added,
+        )
+
+        return artifact_bundle, True
+    else:
+        # We store a reference to the previous file to which the bundle was pointing to.
+        existing_file = existing_artifact_bundle.file
+
+        # In case there is an ArtifactBundle with a specific bundle_id, we want to change its underlying File model
+        # with its corresponding artifact count.
+        existing_artifact_bundle.update(
+            date_added=date_added, file=archive_file, artifact_count=artifact_count
+        )
+
+        # We now delete that file, in order to avoid orphan files in the database.
+        existing_file.delete()
```

Comment:
I believe files should also be deduped by their checksum, no? Maybe we should check that edge-case here to make sure that `existing_artifact_bundle.file != archive_file`?